### PR TITLE
Kad Import

### PR DESCRIPTION
**Pull request by guillaumemichel**
See https://github.com/plprobelab/go-kademlia/issues/120#issuecomment-1714165050 for reference.

Importing basic Kademlia types, as well as the `trie`, `key` and `triert` packages.

### CODE REVIEW

File: kad/key/key256/key256_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/libp2p/go-libdht/kad/key/test"
+)
+
+func TestKey256(t *testing.T) {
+	tester := &test.KeyTester[Key256]{
+		// kt.Key0 is 00000...000
+		Key0: ZeroKey256(),
+
+		// key1 is key0 + 1 (00000...001)
+		Key1: NewKey256([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01}),
```
Comment (dennis-tra) [inline kad/key/key256/key256_test.go:15]:
```suggestion
		Key1: NewKey256(append(make([]byte, 31), 0x01)),
```

File: kad/key/key256/key256_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+		// key2 is key0 + 2 (00000...010)
+		Key2: NewKey256([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02}),
```
Comment (dennis-tra) [inline kad/key/key256/key256_test.go:18]:
```suggestion
		Key2: NewKey256(append(make([]byte, 31), 0x02)),
```

File: kad/key/key256/key256_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+		// key1xor2 is key1 ^ key2 (00000...011)
+		Key1xor2: NewKey256([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03}),
```
Comment (dennis-tra) [inline kad/key/key256/key256_test.go:21]:
```suggestion
		Key1xor2: NewKey256(append(make([]byte, 31), 0x03)),
```

File: kad/key/key256/key256_test.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+		// key100 is key0 with the most significant bit set (10000...000)
+		Key100: NewKey256([]byte{0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}),
+
+		// key010 is key0 with the second most significant bit set (01000...000)
+		Key010: NewKey256([]byte{0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}),
```
Comment (dennis-tra) [inline kad/key/key256/key256_test.go:27]:
```suggestion
		Key010: NewKey256(append([]byte{0x40}, make([]byte, 31)...)),
```

File: kad/key/key256/key256_test.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+		KeyX: NewKey256([]byte{0x23, 0xe4, 0xdd, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}),
```
Comment (dennis-tra) [inline kad/key/key256/key256_test.go:29]:
```suggestion
		KeyX: NewKey256(append([]byte{0x23, 0xe4, 0xdd, 0x03}, make([]byte, 28)...)),
```

**Comment by dennis-tra** [comment]
Awesome, thanks @guillaumemichel! This is how I thought the first step could look like. Love the `go.mod` with just a single test dependency. Let's see how long it'll stay that lean.

Some comments:

- I'd prefer a shallow package hierarchy where the different key types would just live in a single directory, i.e., lift everything from `bistrkey`, `key256`, and `key/test` into the `key` package/folder. `BitStrKey` would then become only `BitStr` or `BitString`. `Key256` would probably stay `Key256` because the name cannot start with a number.
- do we want to lift `kadtest` out of the internal package? We're using things from there in go-libp2p-kad-dht - for example the CtxShort. Probably more @iand might have a better overview?
- Would be great to pull in UCI for automated tests (can be a separate PR)

**Comment by guillaumemichel** [comment]
> I'd prefer a shallow package hierarchy where the different key types would just live in a single directory, i.e., lift everything from bistrkey, key256, and key/test into the key package/folder. BitStrKey would then become only BitStr or BitString. Key256 would probably stay Key256 because the name cannot start with a number.

What are the benefits of a shallow package hierarchy?

The benefit I see for the current hierarchy are that users can choose to import only one type of `Key` and they don't end up importing a large package with many Keys they don't use. I am not against having all implementations in a `key` package, but would like to understand the benefits.

> Would be great to pull in UCI for automated tests (can be a separate PR)

What is the process to pull these?

> There are a couple of interfaces in kad that we don't use anymore afaict. For example, Address and both <X>Protocol interfaces.

In this case I suggest to remove the following types:
* `NodeInfo`
* `Address`
* `Message`
* `Request`
* `Response`
* `RoutingProtocol`
* `RecordProtocol`
* `Record`

We can implement the necessary types in `go-libp2p-kad-dht/v2` and move back some types here once we are happy with them if required.

**Comment by iand** [comment]
My preference is also for shallower package hierarchies. But I think having a package for distinct types of key makes sense. I see an analogy with the crypto or hash package hierarches in the Go standard lib

However the current types names are awkward to work with and leads to stuttering:

 - bitstrkey.BitStrKey
 - key256.Key256

I would prefer:

 - bitstr.Key

key256 is hard to rename but perhaps this would be nicer

 - bit256.Key

**Comment by dennis-tra** [comment]
Hmm I'd still prefer a single key package :D the crypto packages are more sophisticated than what we're doing with our keys but it seems I'm outvoted :)

Fine with me to merge 👍

**Review by dennis-tra — APPROVED** [review]

**Pull request merged**